### PR TITLE
sd-boot: Be silent on regular boots

### PR DESCRIPTION
**Pull request by medhefgo**
We already print an error when this test fails, so no
need to annoy people on a daily basis.

### CODE REVIEW

**Comment by medhefgo** [comment]
Note that theoretically this shouldn't be in production builds because this test is guarded by `#ifndef __OPTIMIZE__`. But sd-boot is always built with -O0, so `__OPTIMIZE__` is not defined.

**Comment by poettering** [comment]
hmm, this doesn't look right... Why is sd-boot built with -O0? that sounds like something to fix

**Comment by medhefgo** [comment]
I just assumed there was a good reason for that and didn't even try changing that. Also, the build file should be changed by someone else then, cause I don't have the confidence (or patience) for that.

**Comment by leipero** [comment]
Just to add, this should help for the situation when FRAMEBUFFER_CONSOLE_DEFERRED_TAKEOVER is used, because when it's used, annoying msg "SHA256 validated" stays till system completely boots up into display manager. Obviously this should not happen in such scenario because it destroys clean silent boot.

**Comment by medhefgo** [comment]
I've changed build file to only disable optimization for debug builds.

**Comment by kyrias** [comment]
Looking in the gummiboot Git history, it's been built with `-O0` since the first Git commit by Kay.  So if anyone would know, hopefully it'd be him?

**Comment by lnicola** [comment]
Looks like CI is failing because of a warning in `menu_run` in `src/boot/efi/boot.c`. The code is a bit convoluted, and the compiler thinks `err` is uninitialized. Maybe it's because of the optimization level change?

@medhefgo Can you fix that warning, e.g.

```diff
diff --git i/src/boot/efi/boot.c w/src/boot/efi/boot.c
index fd0c972951..2cf32892ce 100644
--- i/src/boot/efi/boot.c
+++ w/src/boot/efi/boot.c
@@ -515,7 +515,6 @@ static BOOLEAN menu_run(
         BOOLEAN exit = FALSE;
         BOOLEAN run = TRUE;
         BOOLEAN wait = FALSE;
-        BOOLEAN cleared_screen = FALSE;
 
         graphics_mode(FALSE);
         uefi_call_wrapper(ST->ConIn->Reset, 2, ST->ConIn, FALSE);
@@ -527,16 +526,12 @@ static BOOLEAN menu_run(
 
         if (config->console_mode_change != CONSOLE_MODE_KEEP) {
                 err = console_set_mode(&config->console_mode, config->console_mode_change);
-                if (!EFI_ERROR(err))
-                        cleared_screen = TRUE;
+                if (EFI_ERROR(err)) {
+                        uefi_call_wrapper(ST->ConOut->ClearScreen, 1, ST->ConOut);
+                        Print(L"Error switching console mode to %ld: %r.\r", (UINT64)config->console_mode, err);
+                }
         }
 
-        if (!cleared_screen)
-                uefi_call_wrapper(ST->ConOut->ClearScreen, 1, ST->ConOut);
-
-        if (config->console_mode_change != CONSOLE_MODE_KEEP && EFI_ERROR(err))
-                Print(L"Error switching console mode to %ld: %r.\r", (UINT64)config->console_mode, err);
-
         err = uefi_call_wrapper(ST->ConOut->QueryMode, 4, ST->ConOut, ST->ConOut->Mode->Mode, &x_max, &y_max);
         if (EFI_ERROR(err)) {
                 x_max = 80;
```

**Comment by nanoant** [comment]
I believe the CI is failing because of new `-O2` that does some optimization and now complains. As far I can tell `src/boot/efi/boot.c` is not part of PR, so nothing has changes there.
~~~
FAILED: src/boot/efi/boot.c.o 
cc -c ../src/boot/efi/boot.c -o src/boot/efi/boot.c.o -Wall -Wextra -std=gnu90 -nostdinc -fpic -fshort-wchar -ffreestanding -fno-strict-aliasing -fno-stack-protector -Wsign-compare -Wno-missing-field-initializers -isystem /usr/include/efi -isystem /usr/include/efi/x86_64 -include src/boot/efi/efi_config.h -include version.h -mno-red-zone -mno-sse -mno-mmx -DEFI_FUNCTION_WRAPPER -DGNU_EFI_USE_MS_ABI -Werror -O2
../src/boot/efi/boot.c: In function ‘menu_run’:
../src/boot/efi/boot.c:538:17: error: ‘err’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
                 Print(L"Error switching console mode to %ld: %r.\r", (UINT64)config->console_mode, err);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
~~~

The offending part, that I agree is unnecessarily convoluted is: (`//!` are my comments to the code)
~~~c
        EFI_STATUS err;
        //! ... removed not relevant code
        BOOLEAN cleared_screen = FALSE;

        if (config->console_mode_change != CONSOLE_MODE_KEEP) {
                //! this is the place that err may be initialized to some concrete value
                err = console_set_mode(&config->console_mode, config->console_mode_change);
                if (!EFI_ERROR(err))
                        cleared_screen = TRUE;
        }

        if (!cleared_screen)
                uefi_call_wrapper(ST->ConOut->ClearScreen, 1, ST->ConOut);

        //! in meantime compiler may assume config->console_mode_change is not the same as few lines above, since config points to external thing that can be changed
        if (config->console_mode_change != CONSOLE_MODE_KEEP && EFI_ERROR(err)) //! <- err may be uninitialized at this point
                Print(L"Error switching console mode to %ld: %r.\r", (UINT64)config->console_mode, err);
~~~

To me far more readable shorter thing that is equivalent and should not raise any error is following:
~~~c
        EFI_STATUS err;
        //! ... removed not relevant code
        if (config->console_mode_change != CONSOLE_MODE_KEEP) {
                err = console_set_mode(&config->console_mode, config->console_mode_change);
                if (EFI_ERROR(err)) {
                        uefi_call_wrapper(ST->ConOut->ClearScreen, 1, ST->ConOut);
                        Print(L"Error switching console mode to %ld: %r.\r", (UINT64)config->console_mode, err);
                }        
        }
~~~

File: src/boot/efi/random-seed.c [hunk 1]
```diff
@@ -168,7 +164,6 @@ EFI_STATUS acquire_system_token(VOID **ret, UINTN *ret_size) {
 }
 
 static VOID validate_sha256(void) {
-
 #ifndef __OPTIMIZE__
```
Comment (keszybz) [inline src/boot/efi/random-seed.c]: Please drop this `#ifndef`. We generally don't want to change behaviour based on complication flags. The check should be very quick, so it is not a problem to enable it uncondtitionally.

Comment (medhefgo) [reply src/boot/efi/random-seed.c]: That defeats the purpose of making the build actually optizimed and will reintroduce the message we want to get rid of in the first place. This just shouldn't be in production builds.

Comment (lnicola) [reply src/boot/efi/random-seed.c]: > i mean, the msg is not the problem, we shouldn't validate SHA256 at all if we are running a production build

Please don't run unit tests in production builds.

**Comment by nanoant** [comment]
Thanks to everyone involved 👏

**Pull request merged**